sapi/cli/php_cli_server.c

```diff
@@ -384,24 +381,17 @@ static const char *get_mime_type(const php_cli_server *server, const char *ext,
 PHP_FUNCTION(apache_request_headers) /* {{{ */
 {
 	php_cli_server_client *client;
-	HashTable *headers;
-	zend_string *key;
-	char *value;
 	zval tmp;
 
 	if (zend_parse_parameters_none() == FAILURE) {
 		RETURN_THROWS();
 	}
 
 	client = SG(server_context);
-	headers = &client->request.headers_original_case;
 
-	array_init_size(return_value, zend_hash_num_elements(headers));
-
-	ZEND_HASH_MAP_FOREACH_STR_KEY_PTR(headers, key, value) {
-		ZVAL_STRING(&tmp, value);
-		zend_symtable_update(Z_ARRVAL_P(return_value), key, &tmp);
-	} ZEND_HASH_FOREACH_END();
+	/* Need to copy the HashTable */
+	ZVAL_ARR(&tmp,&client->request.headers_original_case);
+	RETURN_COPY(&tmp);
```

Comment:
Might be wrong about this (especially if memory sanitizers are happy) but doesn't `ZVAL_ARR` discard (or rather not copy) the `IS_TYPE_REFCOUNTED` flag, which would mean that `RETURN_COPY` wouldn't actually increase the refcount (and could potentially mutate the same array). You'd need `ZVAL_COPY_VALUE` to actually copy those flags (without increasing the refcount itself), but I wonder if you need a `tmp` value at all or can't just do `RETURN_COPY(&client->request.headers_original_case);`.

Reply:
Oh, as suspected, my guess was wrong, `IS_ARRAY_EX` contains `IS_TYPE_REFCOUNTED`.